File: src/jl/test/analysis/test_interval_arithmetic.jl

@testset "interval arithmetic" begin
    @test Interval(-1,2) + Interval(2,3) == Interval(1,5)
    @test Interval(-1,2) * Interval(2,3) == Interval(-3,6)
    @test Interval(-1,2) - Interval(2,3) == Interval(-4,0)
    @test Interval(-1,2) / Interval(2,3) == Interval(-0.5, 1.0)
    @test Interval(-1,2) * Interval(3) == Interval(-3,6)
    @test Interval(-1,2) ∪ Interval(2,3) == Interval(-1,3)
    @test Interval(-1,2) ∪ Interval(3,4) == Interval(-1,4)
    @test Interval(-1,2) ^ Interval(2,3) == Interval(0, Inf)
    @test Interval(-1,2) ^ Interval(2) == Interval(0,4)
    @test Interval(-1,2) ^ Interval(3) == Interval(-1,8)


    source_code = """
    a = x + y
    b = a * z
    exp(b)
    """
    syntax_tree = get_syntax_tree_for_str(source_code)
    scoped_tree = get_scoped_tree(syntax_tree)
    cfg_progr_repr = get_cfg_representation(scoped_tree)

    node_to_evaluate = syntax_tree.root_node[3]
    x = Interval(-1,2)
    y = Interval(2,3)
    z = Interval(2)
    result = static_interval_eval(cfg_progr_repr, node_to_evaluate, Dict(:x => x, :y => y, :z => z))
    @test result == exp((x + y) * z)

    source_code = """
    if x < 3
        a = x + y
    else
        a = x - y
    end
    b = a * z
    exp(b)
    """
    syntax_tree = get_syntax_tree_for_str(source_code)
    scoped_tree = get_scoped_tree(syntax_tree)
    cfg_progr_repr = get_cfg_representation(scoped_tree)
    
    node_to_evaluate = syntax_tree.root_node[3]
    result = static_interval_eval(cfg_progr_repr, node_to_evaluate, Dict(:x => x, :y => y, :z => z))
    @test result == exp(((x + y) ∪ (x - y)) * z)


    source_code = """
    if x < 3
        a = x + y()
    else
        a = x - y()
    end
    b = a * z
    exp(b)
    """
    syntax_tree = get_syntax_tree_for_str(source_code)
    scoped_tree = get_scoped_tree(syntax_tree)
    cfg_progr_repr = get_cfg_representation(scoped_tree)
    
    node_to_evaluate = syntax_tree.root_node[3]
    result = static_interval_eval(cfg_progr_repr, node_to_evaluate, Dict(:x => x, :y => y, :z => z))
    @test result == exp(((x + y) ∪ (x - y)) * z)

    source_code = """
    x = 1
    while true
        x = x + 1
    end
    exp(x)
    """
    syntax_tree = get_syntax_tree_for_str(source_code)
    scoped_tree = get_scoped_tree(syntax_tree)
    cfg_progr_repr = get_cfg_representation(scoped_tree)
    
    node_to_evaluate = syntax_tree.root_node[3]
    @test_throws StackOverflowError _static_interval_eval(cfg_progr_repr, node_to_evaluate, Dict{Symbol, Interval}())
    @test Interval(-Inf, Inf) == static_interval_eval(cfg_progr_repr, node_to_evaluate, Dict{Symbol, Interval}())
    
end

File: src/py/test/test_analysis/test_interval_arithmetic.py
import unittest
import sys
sys.path.insert(0, 'src/py') # hack for now

import ast
from ast_utils.utils import *
from ast_utils.preprocess import *
from ast_utils.scoped_tree import get_scoped_tree

from analysis.interval_arithmetic import *

class TestIntervalArithmetic(unittest.TestCase):
    def test_1(self):
        self.assertEqual(add(Interval(-1,2), Interval(2,3)), Interval(1,5))
        self.assertEqual(mul(Interval(-1,2), Interval(2,3)), Interval(-3,6))
        self.assertEqual(sub(Interval(-1,2), Interval(2,3)), Interval(-4,0))
        self.assertEqual(div(Interval(-1,2), Interval(2,3)), Interval(-0.5, 1.0))
        self.assertEqual(mul(Interval(-1,2), Interval(3)), Interval(-3,6))
        self.assertEqual(union(Interval(-1,2), Interval(2,3)), Interval(-1,3))
        self.assertEqual(union(Interval(-1,2), Interval(3,4)), Interval(-1,4))
        self.assertEqual(pow(Interval(-1,2), Interval(2)), Interval(0,4))
        self.assertEqual(pow(Interval(-1,2), Interval(3)), Interval(-1,8))

    def test_2(self):
        source_code = """
a = x + y
b = a * z
exp(b)
        """
        parsed_ast = ast.parse(source_code)
        line_offsets = get_line_offsets_for_str(source_code)
        syntax_tree = preprocess_syntaxtree(parsed_ast, source_code, line_offsets, 0)
        scoped_tree = get_scoped_tree(syntax_tree)

        node_to_evaluate = scoped_tree.root_node.body[2].value
        x = Interval(-1,2)
        y = Interval(2,3)
        z = Interval(2)
        result = static_interval_eval(scoped_tree, node_to_evaluate, {"x": x, "y": y, "z": z})
        self.assertEqual(result, exp(mul(add(x, y), z)))

    
    def test_3(self):
        source_code = """
if x < 3:
    a = x + y
else:
    a = x - y
b = a * z
exp(b)
        """
        parsed_ast = ast.parse(source_code)
        line_offsets = get_line_offsets_for_str(source_code)
        syntax_tree = preprocess_syntaxtree(parsed_ast, source_code, line_offsets, 0)
        scoped_tree = get_scoped_tree(syntax_tree)

        node_to_evaluate = scoped_tree.root_node.body[2].value
        x = Interval(-1,2)
        y = Interval(2,3)
        z = Interval(2)
        result = static_interval_eval(scoped_tree, node_to_evaluate, {"x": x, "y": y, "z": z})
        self.assertEqual(result, exp(mul(union(add(x, y), sub(x, y)), z)))


    def test_4(self):
        source_code = """
if x < 3:
    a = x + y()
else:
    a = x - y()
b = a * z
exp(b)
        """
        parsed_ast = ast.parse(source_code)
        line_offsets = get_line_offsets_for_str(source_code)
        syntax_tree = preprocess_syntaxtree(parsed_ast, source_code, line_offsets, 0)
        scoped_tree = get_scoped_tree(syntax_tree)

        node_to_evaluate = scoped_tree.root_node.body[2].value
        x = Interval(-1,2)
        y = Interval(2,3)
        z = Interval(2)
        result = static_interval_eval(scoped_tree, node_to_evaluate, {"x": x, "y": y, "z": z})
        self.assertEqual(result, exp(mul(union(add(x, y), sub(x, y)), z)))